### PR TITLE
Ignore events from other revisions in cache handler

### PR DESCRIPTION
**Pull request by morepork**
This fixes a bug where ServiceEndpoints from other revisions were being processed and endpoints from them were being created.

Fixes https://github.com/istio/istio/issues/42212

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
I would suggest we fo revision check in `pilot/pkg/config/kube/crdclient/cache_handler.go`
seems should be `createCacheHandler`, it is to filter out all the unrelated CRs

**Review by howardjohn — COMMENTED** [review]
Wonder if this should instead be handled in crdclient onEvent.

I think right now we would also trigger XDS pushes for other configs (like VS) in other revisions even though we ignore them. This would fix that, and also consolidate the "deleted" logic maybe...?

Not sure that ends up being cleaner though or just moves more complexity into crdclient

**Comment by morepork** [comment]
I tried doing it in cache_handler first, but I didn't get it quite right around the initial scan when the istiod pod started up. I'll have another try at getting that version to work

**Comment by morepork** [comment]
@hzxuzhonghu and @howardjohn I've moved the checks into the crdclient code, which is probably better as it means other revisions will be skipped for everything that registers handlers, and should be a performance win for users of multiple revisions.

I don't think it's any more complex than the previous approach. Note that it also adds a revision check to the initial sync.

**Comment by hzxuzhonghu** [comment]
Thanks, will have a look

File: releasenotes/notes/42212.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - 42212
+
+releaseNotes:
+- |
+  **Fixed** ServiceEntries with other revision labels were being processed.
```
Comment (hzxuzhonghu) [inline releasenotes/notes/42212.yaml:9]: not only SE, but also any other CRs

**Review by hzxuzhonghu — APPROVED** [review]

File: releasenotes/notes/42212.yaml (hunk 1 above)
Comment (morepork) [reply releasenotes/notes/42212.yaml:9]: I was unsure how to word this, the only bug I saw was that SEs in other revisions were being processed fully. Processed is perhaps not the best word here.

I didn't try every other resource, but for the ones I did they were I guess "partially" processed, what would happen is that the event would come through from the crd client. Then they would not use the resource from the event directly, instead rebuild from scratch using the cache. The `List` and `Get` functions in the cache did have revision checks so places that used them would end up doing extra work, but not create incorrect configuration.

**Review by deva26 — APPROVED** [review]

File: pilot/pkg/config/kube/crdclient/client_test.go [hunk 2]
```diff
@@ -327,6 +331,75 @@ func TestClient(t *testing.T) {
 	})
 }
 
+// TestClientInitialSyncSkipsOtherRevisions tests that the initial sync skips objects from other
+// revisions.
+func TestClientInitialSyncSkipsOtherRevisions(t *testing.T) {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/client_test.go:336]: getting some flaky tests on this:

```
=== RUN   TestClientInitialSyncSkipsOtherRevisions
2022-12-05T22:08:34.949555Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1alpha2/ReferenceGrant as it is not present     controller=
2022-12-05T22:08:34.949705Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1alpha2/TCPRoute as it is not present   controller=
2022-12-05T22:08:34.949777Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1alpha2/TLSRoute as it is not present   controller=
2022-12-05T22:08:34.949852Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1beta1/GatewayClass as it is not present        controller=
2022-12-05T22:08:34.949901Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1beta1/Gateway as it is not present     controller=
2022-12-05T22:08:34.949933Z     warn    kube    Skipping CRD gateway.networking.k8s.io/v1beta1/HTTPRoute as it is not present   controller=
2022-12-05T22:08:34.950097Z     info    kube    Starting Pilot K8S CRD controller       controller=
2022-12-05T22:08:34.950304Z     info    kube    controller "security.istio.io/v1beta1/AuthorizationPolicy" is syncing...        controller=
2022-12-05T22:08:34.952901Z     info    kube    controller "networking.istio.io/v1alpha3/WorkloadGroup" is syncing...   controller=
2022-12-05T22:08:34.961335Z     info    kube    controller "telemetry.istio.io/v1alpha1/Telemetry" is syncing...        controller=
2022-12-05T22:08:34.978650Z     info    kube    Pilot K8S CRD controller synced in 28.576186ms  controller=
    client_test.go:399: expected configs [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }] do not match those added [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }]
--- FAIL: TestClientInitialSyncSkipsOtherRevisions (0.05s)
FAIL
```

```
==================
WARNING: DATA RACE
Read at 0x00c000d222e8 by goroutine 57:
  reflect.typedmemmove()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/runtime/mbarrier.go:178 +0x0
  reflect.packEface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:123 +0xaf
  reflect.valueInterface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1508 +0x18b
  reflect.Value.Interface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1478 +0xc9
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:753 +0xca
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d222e8 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000d22328 by goroutine 57:
  reflect.packEface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:130 +0xfd
  reflect.valueInterface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1508 +0x18b
  reflect.Value.Interface()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1478 +0xc9
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:753 +0xca
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d22328 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000d22368 by goroutine 57:
  reflect.Value.Len()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1617 +0x247d
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:892 +0x2426
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1db5
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d22368 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c000d22380 by goroutine 57:
  reflect.Value.IsNil()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/reflect/value.go:1552 +0x99
  fmt.getField()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:323 +0x65
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:837 +0x1d89
  fmt.(*pp).printValue()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:896 +0x2425
  fmt.(*pp).printArg()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:743 +0xe17
  fmt.(*pp).doPrintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:1057 +0x499
  fmt.Sprintf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/fmt/print.go:219 +0x67
  testing.(*common).Fatalf()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:954 +0x8a
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:399 +0xe1e
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47

Previous write at 0x00c000d22380 by goroutine 58:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:380 +0x104
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).callHandlers()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:97 +0x282
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*cacheHandler).onEvent()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:76 +0xc27
  istio.io/istio/pilot/pkg/config/kube/crdclient.createCacheHandler.func2.1()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/cache_handler.go:125 +0x53
  istio.io/istio/pkg/queue.(*queueImpl).processNextItem()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:109 +0x5c
  istio.io/istio/pkg/queue.(*queueImpl).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pkg/queue/instance.go:135 +0x20d
  istio.io/istio/pilot/pkg/config/kube/crdclient.(*Client).Run()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client.go:260 +0x344
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions.func3()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0x47

Goroutine 57 (running) created at:
  testing.(*T).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1846 +0x99
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.runTests()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1844 +0x7ec
  testing.(*M).Run()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1726 +0xa84
  istio.io/istio/tests/util/leak.CheckMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/tests/util/leak/check.go:151 +0x3b
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestMain()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/leak_test.go:25 +0x32b
  main.main()
      _testmain.go:59 +0x31f

Goroutine 58 (running) created at:
  istio.io/istio/pilot/pkg/config/kube/crdclient.TestClientInitialSyncSkipsOtherRevisions()
      /usr/local/google/home/howardjohn/go/src/istio.io/istio/pilot/pkg/config/kube/crdclient/client_test.go:386 +0xbe9
  testing.tRunner()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/google/home/howardjohn/sdk/go1.19/src/testing/testing.go:1493 +0x47
==================
    client_test.go:399: expected configs [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }] do not match those added [{{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-0 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-3 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  } {{networking.istio.io/v1alpha3/ServiceEntry  test-service-entry-6 test  map[] map[]  0001-01-01 00:00:00 +0000 UTC [] 0}  }]
    testing.go:1319: race detected during execution of test
--- FAIL: TestClientInitialSyncSkipsOtherRevisions (0.06s)
=== CONT
    testing.go:1319: race detected during execution of test
FAIL
```

**Review by howardjohn — COMMENTED** [review]
Code looks great, I think we just need some small tweak to the test

File: pilot/pkg/config/kube/crdclient/client_test.go (hunk 2 above)
Comment (morepork) [reply pilot/pkg/config/kube/crdclient/client_test.go:336]: Thanks for finding that, that race was incredibly rare on my machine had to run the tests 20,000 times to get it to reliably repro.

I've fixed it by calling `SyncAll()` directly to do the initial bootstrap sync. I tried closing the stop channel and waiting for the store to finish, but that led to another rare flake where the k8s fake client would occasionally send extra add events that the store would then pick up, so calling `SyncAll()` appears to be the safest option.

```
% go test -race -count=20000 ./pilot/pkg/config/kube/crdclient/ -test.run=TestClientInitialSyncSkipsOtherRevisions
ok  	istio.io/istio/pilot/pkg/config/kube/crdclient	75.443s
```

**Comment by howardjohn** [comment]
Nice, fwiw https://github.com/istio/istio/wiki/test%20flakes has some tips
on reproducing

**Comment by morepork** [comment]
@howardjohn - I have resolved the flaky test, would you be able to take another look?

**Review by howardjohn — APPROVED** [review]

**Comment by zirain** [comment]
wait https://github.com/istio/istio/pull/42356

**Comment by morepork** [comment]
> wait #42356

Thanks - those tests passed for me this morning (and locally) so I was unsure what was up in CI

**Pull request merged**